src/compiler/cpucall.jl

```diff
@@ -0,0 +1,319 @@
+const CPUCALL_AREA = "hostcall_area"
+
+const cpucall_areas = Dict{CuContext, Mem.HostBuffer}()
+has_hostcalls(ctx::CuContext) = haskey(cpucall_areas, ctx)
+
+function wait_and_kill_watcher(e::CuEvent, ctx::CuContext)
+    has_hostcalls(ctx) || return
+
+    println("Start the watcher!")
+
+    try
+        while !query(e)
+            handle_cpucall(ctx)
+            yield()
+        end
+    catch e
+        println("Failed $e")
+    end
+
+    println("Killed the watcher!")
+end
+
+cpucall_area_size() = sizeof(UInt8) * 1024 * 1024
+
+@eval @inline cpucall_area() =
+    Base.llvmcall(
+        $("""@hostcall_area = weak externally_initialized global i$(WORD_SIZE) 0
+             define i64 @entry() #0 {
+                 %ptr = load i$(WORD_SIZE), i$(WORD_SIZE)* @hostcall_area, align 8
+                 ret i$(WORD_SIZE) %ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), Ptr{Cvoid}, Tuple{})
+
+
+
+dump_memory(ty=UInt8) = dump_memory{UInt8}(ty)
+dump_memory(ty::Type{T}, size=cpucall_area_size() ÷ sizeof(ty), ctx::CuContext=context()) where {T} = dump_memory(ty, size, ctx)
+function dump_memory(::Type{T}, size::Int64, ctx::CuContext=context()) where {T}
+    ptr    = convert(CuPtr{T}, cpucall_areas[ctx])
+    cuarray = unsafe_wrap(CuArray{T}, ptr, size)
+    println("Dump $cuarray")
+end
+
+
+function reset_cpucall_area!(ctx::CuContext)
+    has_hostcalls(ctx) || return
+
+    println("resetting")
+    ptr    = convert(Ptr{Int64}, cpucall_areas[ctx])
+
+    unsafe_store!(ptr, 0, 1)
+    unsafe_store!(ptr, 0, 2)
+end
+
+
+function create_cpucall_area!(mod::CuModule)
+    flag_ptr = CuGlobal{Ptr{Cvoid}}(mod, CPUCALL_AREA)
+    cpucall_area = get!(cpucall_areas, mod.ctx,
+        Mem.alloc(Mem.Host, cpucall_area_size(), Mem.HOSTALLOC_DEVICEMAP | Mem.HOSTALLOC_WRITECOMBINED))
+    flag_ptr[] = reinterpret(Ptr{Cvoid}, convert(CuPtr{Cvoid}, cpucall_area))
+
+    ptr    = convert(CuPtr{UInt8}, cpucall_area)
+    cuarray = unsafe_wrap(CuArray{UInt8}, ptr, cpucall_area_size())
+    fill!(cuarray, 0)
+
+    return
+end
+
+
+
+time_ms() = round(Int64, time() * 1000)
+
+mutable struct Timer
+    name :: String
+    last_handled :: Int64
+end
+
+const timer = Timer("timer 1", time_ms())
+const timer2 = Timer("timer 2", time_ms())
+
+function update!(timer::Timer)
+    time = time_ms()
+    if time > timer.last_handled + 1
+        println("$(timer.name): It's been a while $(time - timer.last_handled)ms")
+    end
+    timer.last_handled = time
+end
+
+function handle_cpucall(ctx::CuContext)
+    # update!(timer)
+    ptr    = convert(Ptr{Int64}, cpucall_areas[ctx])
+    llvmptr = reinterpret(Core.LLVMPtr{Int64,AS.Global}, ptr)
+    ptr_u8 = convert(Ptr{UInt8}, cpucall_areas[ctx])
+
+    v = atomic_cas!(llvmptr, CPU_CALL, CPU_HANDLING)
+
+    if v != CPU_CALL
+        return
+    end
+
+    # this code is very fragile
```

Comment:
That won't just be fragile, right? https://stackoverflow.com/questions/23193151/atomic-operations-in-cuda-kernels-on-mapped-pinned-host-memory-to-do-or-not-to
Maybe some kind of ring buffer where the CPU just passively reads would be safer? That may only require atomicity from the GPU's side.